Invalid Charge when using suffix currency

So I think the shop signs does not play well with suffix currencies, because if I try to use signs with suffix currencies, it throws an error.

### Information

**Full output of `/ess version`**:

```
Server version: 1.14.4-R0.1-SNAPSHOT git-Paper-243 (MC: 1.14.4)
EssentialsX version: 2.17.2.77
LuckPerms version: 5.0.72
Vault version: 1.7.2-b107
PlaceholderAPI version: 2.10.6
EssentialsXChat version: 2.17.2.77
EssentialsXAntiBuild version: 2.17.2.77
```

### Details

**Steps to reproduce**  
1. Enable buy/sell sign and set ```currency-symbol-suffix``` to true
2. Place a buy/sell sign of any item using currency cost, e.g
```
[Sell]
1
stone
$1
```
This will change to the following when "Done" is clicked:
```
[Sell]
1
stone
1$
```
3. Use the sign

**Expected behavior**  
For the example above, it should tell you in chat: "1$ has been added to your account"

**Actual behavior**
An error was shown: "Error: Invalid charge"

**Errors in console**
Activating debug mode, this is the error:
```
[14:27:05 INFO]: [Essentials] Error calling the command /\ sign: Sell

com.earth2me.essentials.signs.SignException: §4Invalid charge.

at com.earth2me.essentials.signs.EssentialsSign.getTrade(EssentialsSign.java:417) ~[?:?]

at com.earth2me.essentials.signs.EssentialsSign.getTrade(EssentialsSign.java:404) ~[?:?]

at com.earth2me.essentials.signs.EssentialsSign.getTrade(EssentialsSign.java:400) ~[?:?]

at com.earth2me.essentials.signs.SignSell.onSignInteract(SignSell.java:29) ~[?:?]

at com.earth2me.essentials.signs.EssentialsSign.onSignInteract(EssentialsSign.java:117) ~[?:?]

at com.earth2me.essentials.signs.SignPlayerListener.onSignPlayerInteract(SignPlayerListener.java:61) ~[?:?]

at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor51.execute(Unknown Source) ~[?:?]

at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69) ~[patched_1.14.4.jar:git-Paper-243]

at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80) ~[patched_1.14.4.jar:git-Paper-243]

at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[patched_1.14.4.jar:git-Paper-243]

at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:545) ~[patched_1.14.4.jar:git-Paper-243]

at org.bukkit.craftbukkit.v1_14_R1.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:454) ~[patched_1.14.4.jar:git-Paper-243]

...
```

So looking at the stack trace and the code, I discovered the probable culprit:
```java
    protected final BigDecimal getMoney(final String line) throws SignException {
        final boolean isMoney = line.matches("^[^0-9-\\.][\\.0-9]+$");
        return isMoney ? getBigDecimalPositive(line.substring(1)) : null;
    }
```
The regex only matches prefix currencies written on the sign:
![image](https://user-images.githubusercontent.com/18462068/81502233-e9cf8680-9306-11ea-8e5a-edbb3d2026a4.png)

This results in any suffix currencies returning ```null```. The code in ```getTrade``` then tries to parse it as either usage cost, item cost or xp cost, which obviously will always fail, so any suffix currencies won't be able to use the sign.

I don't know whether this impacts all signs with cost, but I think it should be.

That still didn't work.
This is the error it throws when using a prefix currency ($1):
```
com.earth2me.essentials.signs.SignException: null
	at com.earth2me.essentials.signs.EssentialsSign.getBigDecimal(EssentialsSign.java:396) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.getBigDecimalPositive(EssentialsSign.java:385) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.getMoney(EssentialsSign.java:381) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.getTrade(EssentialsSign.java:414) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.getTrade(EssentialsSign.java:405) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.getTrade(EssentialsSign.java:401) ~[?:?]
	at com.earth2me.essentials.signs.SignSell.onSignInteract(SignSell.java:29) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.onSignInteract(EssentialsSign.java:117) ~[?:?]
	at com.earth2me.essentials.signs.SignPlayerListener.onSignPlayerInteract(SignPlayerListener.java:61) ~[?:?]
	at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor51.execute(Unknown Source) ~[?:?]
	at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69) ~[patched_1.14.4.jar:git-Paper-243]
...
Caused by: java.lang.NumberFormatException
	at java.math.BigDecimal.<init>(Unknown Source) ~[?:?]
	at java.math.BigDecimal.<init>(Unknown Source) ~[?:?]
	at java.math.BigDecimal.<init>(Unknown Source) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.getBigDecimal(EssentialsSign.java:394) ~[?:?]
	... 29 more
Caused by: java.lang.ArrayIndexOutOfBoundsException
```
and the suffix currency (1$):
```
[16:05:51] [Server thread/INFO]: [Essentials] Error calling the command /\ sign: Sell
com.earth2me.essentials.signs.SignException: §4Invalid charge.
	at com.earth2me.essentials.signs.EssentialsSign.getTrade(EssentialsSign.java:418) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.getTrade(EssentialsSign.java:405) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.getTrade(EssentialsSign.java:401) ~[?:?]
	at com.earth2me.essentials.signs.SignSell.onSignInteract(SignSell.java:29) ~[?:?]
	at com.earth2me.essentials.signs.EssentialsSign.onSignInteract(EssentialsSign.java:117) ~[?:?]
	at com.earth2me.essentials.signs.SignPlayerListener.onSignPlayerInteract(SignPlayerListener.java:61) ~[?:?]
	at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor51.execute(Unknown Source) ~[?:?]
	at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69) ~[patched_1.14.4.jar:git-Paper-243]
...
```
This is how I set up the signs:
![2020-05-10_23 12 42](https://user-images.githubusercontent.com/18462068/81504507-e68ec780-9313-11ea-9a34-06d184163120.png)

--- a/Essentials/src/com/earth2me/essentials/Kit.java
+++ b/Essentials/src/com/earth2me/essentials/Kit.java
@@ -7,6 +7,7 @@ import com.earth2me.essentials.textreader.IText;
 import com.earth2me.essentials.textreader.KeywordReplacer;
 import com.earth2me.essentials.textreader.SimpleTextInput;
 import com.earth2me.essentials.utils.DateUtil;
+import com.earth2me.essentials.utils.NumberUtil;
 import net.ess3.api.IEssentials;
 import net.ess3.api.events.KitClaimEvent;
 
@@ -160,11 +161,10 @@ public class Kit {
             boolean spew = false;
             final boolean allowUnsafe = ess.getSettings().allowUnsafeEnchantments();
             final boolean currencyIsSuffix = ess.getSettings().isCurrencySymbolSuffixed();
-            final String currencySymbol = ess.getSettings().getCurrencySymbol();
             List<ItemStack> itemList = new ArrayList<>();
             for (String kitItem : output.getLines()) {
                 if (!currencyIsSuffix ? kitItem.startsWith(ess.getSettings().getCurrencySymbol()) : kitItem.endsWith(ess.getSettings().getCurrencySymbol())) {
-                    final String valueString = currencyIsSuffix ? kitItem.substring(0, currencySymbol.length()) : kitItem.substring(currencySymbol.length());
+                    final String valueString = NumberUtil.sanitizeCurrencyString(kitItem, ess);
                     BigDecimal value = new BigDecimal(valueString.trim());
                     Trade t = new Trade(value, ess);
                     t.pay(user, OverflowType.DROP);

--- a/Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java
+++ b/Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java
@@ -375,9 +375,9 @@ public class EssentialsSign {
         return stack;
     }
 
-    protected final BigDecimal getMoney(final String line) throws SignException {
-        final boolean isMoney = line.matches("^[^0-9-\\.][\\.0-9]+$");
-        return isMoney ? getBigDecimalPositive(line.substring(1)) : null;
+    protected final BigDecimal getMoney(final String line, final IEssentials ess) throws SignException {
+        final boolean isMoney = line.matches("^[^0-9-\\.]?[\\.0-9]+[^0-9-\\.]?$");
+        return isMoney ? getBigDecimalPositive(NumberUtil.sanitizeCurrencyString(line, ess)) : null;
     }
 
     protected final BigDecimal getBigDecimalPositive(final String line) throws SignException {
@@ -410,7 +410,7 @@ public class EssentialsSign {
             return new Trade(signName.toLowerCase(Locale.ENGLISH) + "sign", ess);
         }
 
-        final BigDecimal money = getMoney(line);
+        final BigDecimal money = getMoney(line, ess);
         if (money == null) {
             final String[] split = line.split("[ :]+", 2);
             if (split.length != 2) {

--- a/Essentials/src/com/earth2me/essentials/signs/SignTrade.java
+++ b/Essentials/src/com/earth2me/essentials/signs/SignTrade.java
@@ -149,7 +149,7 @@ public class SignTrade extends EssentialsSign {
         final String[] split = line.split("[ :]+");
 
         if (split.length == 1 && !amountNeeded) {
-            final BigDecimal money = getMoney(split[0]);
+            final BigDecimal money = getMoney(split[0], ess);
             if (money != null) {
                 if (NumberUtil.shortCurrency(money, ess).length() * 2 > 15) {
                     throw new SignException("Line can be too long!");
@@ -160,7 +160,7 @@ public class SignTrade extends EssentialsSign {
         }
 
         if (split.length == 2 && amountNeeded) {
-            final BigDecimal money = getMoney(split[0]);
+            final BigDecimal money = getMoney(split[0], ess);
             BigDecimal amount = getBigDecimalPositive(split[1]);
             if (money != null && amount != null) {
                 amount = amount.subtract(amount.remainder(money));
@@ -218,7 +218,7 @@ public class SignTrade extends EssentialsSign {
 
         if (split.length == 2) {
             try {
-                final BigDecimal money = getMoney(split[0]);
+                final BigDecimal money = getMoney(split[0], ess);
                 final BigDecimal amount = notEmpty ? getBigDecimalPositive(split[1]) : getBigDecimal(split[1]);
                 if (money != null && amount != null) {
                     return new Trade(amountType == AmountType.COST ? money : amount, ess);
@@ -317,7 +317,7 @@ public class SignTrade extends EssentialsSign {
         final String[] split = line.split("[ :]+");
 
         if (split.length == 2) {
-            final BigDecimal money = getMoney(split[0]);
+            final BigDecimal money = getMoney(split[0], ess);
             final BigDecimal amount = getBigDecimal(split[1]);
             if (money != null && amount != null) {
                 final String newline = NumberUtil.shortCurrency(money, ess) + ":" + NumberUtil.shortCurrency(value, ess).substring(1);

--- a/Essentials/src/com/earth2me/essentials/utils/NumberUtil.java
+++ b/Essentials/src/com/earth2me/essentials/utils/NumberUtil.java
@@ -88,6 +88,15 @@ public class NumberUtil {
         return sign + tl("currency", ess.getSettings().getCurrencySymbol(), currency);
     }
 
+    public static String sanitizeCurrencyString(final String input, final IEssentials ess) {
+        String symbol = ess.getSettings().getCurrencySymbol();
+        boolean suffix = ess.getSettings().isCurrencySymbolSuffixed();
+        if (input.contains(symbol)) {
+            return suffix ? input.substring(0, input.indexOf(symbol)) : input.substring(symbol.length());
+        }
+        return input;
+    }
+
     public static boolean isInt(final String sInt) {
         try {
             Integer.parseInt(sInt);

